src/frontend/src/handler/query.rs

```diff
@@ -54,17 +56,28 @@ pub async fn handle_query(
     };
     debug!("query_mode:{:?}", query_mode);
 
+    let (shutdown_tx, shutdown_rx) = oneshot::channel();
     let (data_stream, pg_descs) = match query_mode {
         QueryMode::Local => local_execute(context, bound)?,
-        QueryMode::Distributed => distribute_execute(context, bound).await?,
+        // Local mode do not support cancel tasks.
+        QueryMode::Distributed => distribute_execute(context, bound, shutdown_tx).await?,
     };
 
+    let mut data_stream = data_stream.take_until(shutdown_rx);
     let mut rows = vec![];
     #[for_await]
-    for chunk in data_stream {
+    for chunk in &mut data_stream {
         rows.extend(to_pg_rows(chunk?, format));
     }
 
+    // Check whether error happen, if yes, returned.
+    let execution_ret = data_stream.take_result();
+    if let Some(ret) = execution_ret {
+        return Err(ret
+            .expect("The shutdown message receiver should not fail")
+            .into());
+    }
```

Comment:
I choose `take_until` + `take_result` to early terminate a Stream, and checked correspoding trigger error.